In [40]:
import pandas as pd

In [41]:
df = pd.read_csv('A8-E2_Groceries.csv')

In [42]:
df

Member_number        Date        itemDescription
0               1808  21-07-2015         tropical fruit
1               2552  05-01-2015             whole milk
2               2300  19-09-2015              pip fruit
3               1187  12-12-2015       other vegetables
4               3037  01-02-2015             whole milk
...              ...         ...                    ...
38760           4471  08-10-2014          sliced cheese
38761           2022  23-02-2014                  candy
38762           1097  16-04-2014               cake bar
38763           1510  03-12-2014  fruit/vegetable juice
38764           1521  26-12-2014               cat food

[38765 rows x 3 columns]

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38765 entries, 0 to 38764
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Member_number    38765 non-null  int64 
 1   Date             38765 non-null  object
 2   itemDescription  38765 non-null  object
dtypes: int64(1), object(2)
memory usage: 908.7+ KB


In [44]:
df[['Date', 'Member_number', 'itemDescription']].value_counts()

Date        Member_number  itemDescription
11-03-2015  1994           whole milk         4
23-12-2015  3679           domestic eggs      3
10-01-2015  4981           margarine          3
10-05-2015  2294           rolls/buns         3
28-06-2015  1370           whole milk         3
                                             ..
11-04-2014  2045           yogurt             1
            2151           shopping bags      1
                           tropical fruit     1
            2274           beef               1
31-10-2015  4934           curd               1
Name: count, Length: 38006, dtype: int64

In [45]:
df['itemDescription'].value_counts()

itemDescription
whole milk               2502
other vegetables         1898
rolls/buns               1716
soda                     1514
yogurt                   1334
                         ... 
rubbing alcohol             5
bags                        4
baby cosmetics              3
kitchen utensil             1
preservation products       1
Name: count, Length: 167, dtype: int64

In [46]:
df['Date'].value_counts()

Date
21-01-2015    96
21-07-2015    93
29-11-2015    92
08-08-2015    92
30-04-2015    91
              ..
29-06-2014    26
04-07-2014    24
16-03-2015    23
17-03-2015    23
01-09-2015    22
Name: count, Length: 728, dtype: int64

In [47]:
df['Member_number'].value_counts()

Member_number
3180    36
3737    33
3050    33
2051    33
2625    31
        ..
2503     2
3301     2
1775     2
3723     2
2417     2
Name: count, Length: 3898, dtype: int64

Total unique members is 3898 and total dates is span between 2014 start to end of 2015 i.e 728 days and there is total 167 unique items. 

In [48]:
df[['Date', 'Member_number']].value_counts()

Date        Member_number
12-07-2015  1780             11
06-08-2015  3741              9
03-05-2015  1404              9
29-06-2015  4939              9
28-08-2015  4207              9
                             ..
13-05-2014  1666              2
            2100              2
            2144              2
            2218              2
31-10-2015  4934              2
Name: count, Length: 14963, dtype: int64

On single day means at that moment how many together things as the member bought.
Means there are total 14963 unique transactions.

In [49]:
df[(df['Member_number'] == 1780) & (df['Date']== '12-07-2015')] # This items are bought together.

Member_number        Date   itemDescription
2180            1780  12-07-2015         pip fruit
6706            1780  12-07-2015              curd
7557            1780  12-07-2015     specialty bar
7911            1780  12-07-2015         beverages
8513            1780  12-07-2015           sausage
18329           1780  12-07-2015       soft cheese
22855           1780  12-07-2015        rolls/buns
23706           1780  12-07-2015            yogurt
24060           1780  12-07-2015  other vegetables
24662           1780  12-07-2015              curd
32299           1780  12-07-2015        newspapers

We need to create a new dataframe with columns "Transaction" and "ItemsBought" there will be 14963 rows

In [ ]:
# to create a df with 

In [50]:
df['Date2'] = pd.to_datetime(df['Date'], format='%d-%m-%Y')

In [51]:
print(max(df['Date2']))
print(min(df['Date2'])) # total 728 unique dates.

2015-12-30 00:00:00
2014-01-01 00:00:00


In [52]:
# sorting the dataframe by date
df.sort_values(by=['Date2','Member_number'], inplace=True)
df[['Member_number','Date','itemDescription']]

Member_number        Date   itemDescription
10065           1249  01-01-2014      citrus fruit
26214           1249  01-01-2014            coffee
15338           1381  01-01-2014              curd
31487           1381  01-01-2014              soda
11134           1440  01-01-2014  other vegetables
...              ...         ...               ...
17777           4058  30-12-2015      cream cheese
557             4565  30-12-2015       canned beer
16706           4565  30-12-2015       canned beer
1134            4863  30-12-2015           dessert
17283           4863  30-12-2015              curd

[38765 rows x 3 columns]

In [53]:
df["Transaction"] = "T" + (df.groupby(["Member_number", "Date2"]).ngroup() + 1).astype(str)

In [54]:
df['Transaction'].value_counts()

Transaction
T2869     11
T845       9
T4424      9
T6289      9
T14757     9
          ..
T11550     2
T12106     2
T12974     2
T13092     2
T14497     2
Name: count, Length: 14963, dtype: int64

In [55]:
df[['Transaction', 'itemDescription']].nunique()

Transaction        14963
itemDescription      167
dtype: int64

In [56]:
new_df = df.groupby("Transaction")["itemDescription"].apply(lambda x: ", ".join(x)).reset_index()

# Renaming columns
new_df.columns = ["Transaction", "Items"]

# Display the result
print(new_df)

      Transaction                                              Items
0              T1                    whole milk, pastry, salty snack
1             T10                                  frankfurter, curd
2            T100          root vegetables, long life bakery product
3           T1000  photo/film, whole milk, yogurt, whole milk, wh...
4          T10000                 newspapers, rolls/buns, newspapers
...           ...                                                ...
14958       T9995       shopping bags, newspapers, curd, hard cheese
14959       T9996                             bottled water, tidbits
14960       T9997   specialty bar, candy, root vegetables, ice cream
14961       T9998                                 pork, frozen meals
14962       T9999                      canned beer, other vegetables

[14963 rows x 2 columns]


In [69]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

# Load the dataset (assuming it's stored as a CSV file)
# new_df = pd.read_csv("transactions.csv")  # Replace with your actual file

# Convert 'Items' column into a list of lists (transactions format)
transactions = new_df["Items"].apply(lambda x: x.split(", ")).tolist()

# Convert transactions into a one-hot encoded DataFrame
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
new_df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

# Apply the Apriori algorithm to find frequent itemsets
frequent_itemsets = apriori(new_df_encoded, min_support=0.01, use_colnames=True)

# Generate association rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=0.2)

# Display results
print("Frequent Itemsets:\n", frequent_itemsets)
print("\nAssociation Rules:\n", rules)


Frequent Itemsets:
      support                        itemsets
0   0.021386                      (UHT-milk)
1   0.033950                          (beef)
2   0.021787                       (berries)
3   0.016574                     (beverages)
4   0.045312                  (bottled beer)
..       ...                             ...
64  0.010559  (rolls/buns, other vegetables)
65  0.014837  (whole milk, other vegetables)
66  0.013968        (rolls/buns, whole milk)
67  0.011629              (whole milk, soda)
68  0.011161            (whole milk, yogurt)

[69 rows x 2 columns]

Association Rules:
           antecedents         consequents  antecedent support  \
0        (rolls/buns)  (other vegetables)            0.110005   
1  (other vegetables)        (rolls/buns)            0.122101   
2        (whole milk)  (other vegetables)            0.157923   
3  (other vegetables)        (whole milk)            0.122101   
4        (rolls/buns)        (whole milk)            0.110005   
5     

In [70]:
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])


          antecedents         consequents   support  confidence      lift
0        (rolls/buns)  (other vegetables)  0.010559    0.095990  0.786154
1  (other vegetables)        (rolls/buns)  0.010559    0.086481  0.786154
2        (whole milk)  (other vegetables)  0.014837    0.093948  0.769430
3  (other vegetables)        (whole milk)  0.014837    0.121511  0.769430
4        (rolls/buns)        (whole milk)  0.013968    0.126974  0.804028
5        (whole milk)        (rolls/buns)  0.013968    0.088447  0.804028
6        (whole milk)              (soda)  0.011629    0.073635  0.758296
7              (soda)        (whole milk)  0.011629    0.119752  0.758296
8        (whole milk)            (yogurt)  0.011161    0.070673  0.822940
9            (yogurt)        (whole milk)  0.011161    0.129961  0.822940


In [ ]:
# rules = rules[rules['confidence'] > 0.2]  # Keep only strong rules


In [ ]:
# print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])


Empty DataFrame
Columns: [antecedents, consequents, support, confidence, lift]
Index: []
